### DISCUSSION TITLE
Validation rules for email addresses

### DISCUSSION THREAD

**Discussion by craig-ryan-sn** [swellstores/community#511] [Help] [answered]
What is the RE that swell applies to email addresses? I have an address containing an exclamation point (!) and this fails, is it possible to encode the address so that I can store in swell?

**Comment by ericingram** [comment]
I don't think (!) characters are allowed in email addresses in general

**Reply by surenw-serenade** [reply]
Hi @ericingram ,

We have an example in production of an email in the following format (real data has been obfuscated) :

`joe.o'connor+aaa@mydomain.com`

The user has been able to register an email with this format and thus register with our platform and verify their account as they can receive emails at this address.

Email providers support this format and the documentation in the articles below suggest it is supported as a part of the spec:

https://en.wikipedia.org/wiki/Email_address#Local-part
https://stackoverflow.com/questions/2049502/what-characters-are-allowed-in-an-email-address
https://help.salesforce.com/s/articleView?id=sf.pardot_emails_allowed_characters.htm&type=5

We are having trouble migrating a couple of production customers to Swell due to this issue.

Is there anything you can do to expedite a fix?

**Comment by logeshswell** [comment, marked as answer]
I raised a feature request to add support for (!) character in email validation.

**Reply by logeshswell** [reply]
I will get back to you if I get some updates.

**Reply by ericingram** [reply]
Odd, I found one source that said it isn't allows but some other sources that says it is. We'll look into it further.

**Reply by surenw-serenade** [reply]
@ericingram @logeshswell Thanks for looking into this, really appreciate it.

The feature request isn't just for (!). It would be great to cover all special characters that are a part of the spec or supported by email providers rather than just (!). The example I gave uses an apostrophe (').

I was wondering if you could provide an indicative timeframe for when this will be resolved? As I mentioned, this is affecting our production migration to Swell.

**Reply by logeshswell** [reply]
Sure, I updated the feature request to support all applicable special characters in email validation.
I don't have an ETA at the moment, but I will keep you posted if there's any updates available.

**Reply by logeshswell** [reply]
To unblock you from production migration, I would recommend you to pass `$migrate: true` in the request payload to skip all validations in the corresponding models.

**Comment by steven-k1** [comment]
An improvement has been made for this and special characters are now supported in email validation
